internal/gps/source_cache_bolt_encode.go

```diff
@@ -0,0 +1,439 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gps
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/boltdb/bolt"
+	"github.com/golang/dep/internal/gps/pkgtree"
+	"github.com/pkg/errors"
+)
+
+// cacheEncodeUnpairedVersion returns an encoded UnpairedVersion.
+func cacheEncodeUnpairedVersion(uv UnpairedVersion) ([]byte, error) {
+	var pre string
+	switch uv.Type() {
+	case IsBranch:
+		if uv.(branchVersion).isDefault {
+			pre = "defaultBranch:"
+		} else {
+			pre = "branch:"
+		}
+	case IsSemver, IsVersion:
+		pre = "ver:"
+	default:
+		return nil, fmt.Errorf("unrecognized version type: %d", uv.Type())
+	}
+	return []byte(pre + uv.String()), nil
+}
+
+// cacheDecodeUnpairedVersion decodes and returns a new UnpairedVersion.
+func cacheDecodeUnpairedVersion(b []byte) (UnpairedVersion, error) {
+	const br, dbr, ver = "branch:", "defaultBranch:", "ver:"
+	s := string(b)
+	switch {
+	case strings.HasPrefix(s, br):
+		return NewBranch(strings.TrimPrefix(s, br)), nil
+	case strings.HasPrefix(s, dbr):
+		return newDefaultBranch(strings.TrimPrefix(s, dbr)), nil
+	case strings.HasPrefix(s, ver):
+		return NewVersion(strings.TrimPrefix(s, ver)), nil
+	default:
+		return nil, fmt.Errorf("unrecognized prefix: %s", s)
+	}
+}
+
+// cacheDecodeProjectProperties returns a new ProjectRoot and ProjectProperties with the
+// data encoded in the key/value pair.
+func cacheDecodeProjectProperties(k, v []byte) (ProjectRoot, ProjectProperties, error) {
+	var pp ProjectProperties
+	ks := strings.SplitN(string(k), ",", 2)
+	ip := ProjectRoot(ks[0])
+	if len(ks) > 1 {
+		pp.Source = ks[1]
+	}
+	if len(v) == 0 {
+		pp.Constraint = Any()
+	} else {
+		const br, dbr, ver, rev = "branch:", "defaultBranch:", "ver:", "rev:"
+		vs := string(v)
+		switch {
+		case strings.HasPrefix(vs, br):
+			pp.Constraint = NewBranch(strings.TrimPrefix(vs, br))
+
+		case strings.HasPrefix(vs, dbr):
+			pp.Constraint = newDefaultBranch(strings.TrimPrefix(vs, dbr))
+
+		case strings.HasPrefix(vs, ver):
+			vs = strings.TrimPrefix(vs, ver)
+			if c, err := NewSemverConstraint(vs); err != nil {
+				pp.Constraint = NewVersion(vs)
+			} else {
+				pp.Constraint = c
+			}
+
+		case strings.HasPrefix(vs, rev):
+			pp.Constraint = Revision(strings.TrimPrefix(vs, rev))
+
+		default:
+			return "", ProjectProperties{}, fmt.Errorf("unrecognized prefix: %s", vs)
+		}
+	}
+
+	return ip, pp, nil
+}
+
+// cacheEncodeProjectProperties returns a key/value pair containing the encoded
+// ProjectRoot and ProjectProperties.
+func cacheEncodeProjectProperties(ip ProjectRoot, pp ProjectProperties) ([]byte, []byte, error) {
```

Comment:
Encoding scheme was my first question: https://github.com/golang/dep/issues/431#issuecomment-306786973 ;-)

There is no reason to hand-roll this way other than that it was a simple first step, and that dep's manifest/lock serialization could be used as a reference, so I could be confident about correctness.

> really doing this right may entail pulling up the semver lib directly into dep and modifying it so that we can access its internal data. if it comes to that, we can do it...

We certainly need to access internal data, although we could get away with some other means of exporting the fields - we can always convert into an equivalent type with struct tags or custom serialization methods.

Reply:
> Encoding scheme was my first question: #431 (comment) ;-)

crap, it totally was. i'm sorry, i missed that. but, as you say:

> There is no reason to hand-roll this way other than that it was a simple first step, and that dep's manifest/lock serialization could be used as a reference, so I could be confident about correctness.

i entirely agree. using primitives that we understand and can easily reason about as a first step to getting it working gives us a strong foundation to work from. 👍 

i was just trying approach cautiously in the event that you were quite attached to the hand-rolled scheme, but it doesn't seem you are 🦄 💯 👍 😄 

> We certainly need to access internal data, although we could get away with some other means of exporting the fields - we can always convert into an equivalent type with struct tags or custom serialization methods.

we'd still have to convert _back_ to the original types, though. and there's no mechanism that the semver lib provides for generating objects, apart from its string parser.

Reply:
> we'd still have to convert back to the original types, though. and there's no mechanism that the semver lib provides for generating objects, apart from its string parser.

Yeah the public lib would still have to be augmented, but then we could avoid an internal fork. Any chance that some sort of strongly typed constructor would be generally useful as an addition to the lib?

Reply:
Maybe? At the point where we have a constructor that's exposing all the fields, though, it starts seeming to me like it might just be better to export the fields. i feel like we've started exploring that path in the past, but i can't entirely recall.

Reply:
Just poking around the code I see a few consequences of exporting the fields:
- Making them mutable means the internal semver cache has to hand out copies, not immutable originals
- Exporting fields will make some 'getter' methods obsolete and also clash with their names (e.g. `Version.Major()`), so this is probably API breaking

Reply:
confirmed, i'm expecting both of those to be the case :)

though on the first one, i converted the semver lib a while ago to use values instead of pointers (with the goal of avoiding heap allocating them). possibly an unnecessary microoptimization, but they're generally of sufficiently small size that it may make sense. anyway, if we continue with that, we can just return the dereferenced pointer (assuming the serialization lib operates on a pointer)